File: All DSA C++ Files/java/traversal.java
// In-Order Traversal
class TreeNode {
    int data;
    TreeNode left, right;

    TreeNode(int data) {
        this.data = data;
        left = right = null;
    }
}

public class InOrderTraversal {
    public static void inOrder(TreeNode root) {
        if (root != null) {
            inOrder(root.left);
            System.out.print(root.data + " ");
            inOrder(root.right);
        }
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        System.out.print("In-Order Traversal: ");
        inOrder(root);
        System.out.println();
    }
}


// Pre-Order Traversal
class PreOrderTraversal {
    public static void preOrder(TreeNode root) {
        if (root != null) {
            System.out.print(root.data + " ");
            preOrder(root.left);
            preOrder(root.right);
        }
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        System.out.print("Pre-Order Traversal: ");
        preOrder(root);
        System.out.println();
    }
}

// Post-Order Traversal
class PostOrderTraversal {
    public static void postOrder(TreeNode root) {
        if (root != null) {
            postOrder(root.left);
            postOrder(root.right);
            System.out.print(root.data + " ");
        }
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        System.out.print("Post-Order Traversal: ");
        postOrder(root);
        System.out.println();
    }
}


// DFS Traversal
import java.util.*;

public class DFSTraversal {
    public static void dfs(Map<Integer, List<Integer>> graph, int start, Set<Integer> visited) {
        visited.add(start);
        System.out.print(start + " ");
        for (int neighbor : graph.getOrDefault(start, new ArrayList<>())) {
            if (!visited.contains(neighbor)) {
                dfs(graph, neighbor, visited);
            }
        }
    }

    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(1, 2));
        graph.put(1, Arrays.asList(0, 3, 4));
        graph.put(2, Arrays.asList(0, 4));
        graph.put(3, Arrays.asList(1, 5));
        graph.put(4, Arrays.asList(1, 2, 5));
        graph.put(5, Arrays.asList(3, 4));

        System.out.print("DFS Traversal: ");
        dfs(graph, 0, new HashSet<>());
        System.out.println();
    }
}


// Level-Order Traversal
import java.util.*;

public class LevelOrderTraversal {
    public static void levelOrder(TreeNode root) {
        if (root == null) return;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            System.out.print(node.data + " ");

            if (node.left != null) queue.add(node.left);
            if (node.right != null) queue.add(node.right);
        }
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        System.out.print("Level-Order Traversal: ");
        levelOrder(root);
        System.out.println();
    }
}


// Backtracking Traversal: N-Queens Problem
import java.util.*;

public class NQueens {
    public static boolean isSafe(int[][] board, int row, int col, int n) {
        for (int i = 0; i < col; i++) if (board[row][i] == 1) return false;
        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) if (board[i][j] == 1) return false;
        for (int i = row, j = col; i < n && j >= 0; i++, j--) if (board[i][j] == 1) return false;
        return true;
    }

    public static boolean solveNQueens(int[][] board, int col, int n) {
        if (col >= n) return true;

        for (int i = 0; i < n; i++) {
            if (isSafe(board, i, col, n)) {
                board[i][col] = 1;
                if (solveNQueens(board, col + 1, n)) return true;
                board[i][col] = 0; // Backtrack
            }
        }
        return false;
    }

    public static void printBoard(int[][] board, int n) {
        for (int[] row : board) {
            for (int cell : row) {
                System.out.print((cell == 1 ? "Q " : ". "));
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        int n = 8;
        int[][] board = new int[n][n];
        if (solveNQueens(board, 0, n)) {
            System.out.println("N-Queens Solution:");
            printBoard(board, n);
        } else {
            System.out.println("No solution exists");
        }
    }
}



// A* Search Algorithm
import java.util.*;

public class AStarSearch {
    static class Node implements Comparable<Node> {
        int x, y, cost, heuristic;

        Node(int x, int y, int cost, int heuristic) {
            this.x = x;
            this.y = y;
            this.cost = cost;
            this.heuristic = heuristic;
        }

        @Override
        public int compareTo(Node other) {
            return (this.cost + this.heuristic) - (other.cost + other.heuristic);
        }
    }

    public static List<int[]> aStar(int[][] grid, int[] start, int[] goal) {
        PriorityQueue<Node> openSet = new PriorityQueue<>();
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        Map<int[], int[]> cameFrom = new HashMap<>();

        openSet.add(new Node(start[0], start[1], 0, heuristic(start, goal)));

        while (!openSet.isEmpty()) {
            Node current = openSet.poll();
            if (current.x == goal[0] && current.y == goal[1]) {
                return reconstructPath(cameFrom, start, goal);
            }

            if (visited[current.x][current.y]) continue;
            visited[current.x][current.y] = true;

            for (int[] direction : new int[][] {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}) {
                int nx = current.x + direction[0], ny = current.y + direction[1];
                if (isValid(grid, nx, ny) && !visited[nx][ny]) {
                    int newCost = current.cost + 1;
                    openSet.add(new Node(nx, ny, newCost, heuristic(new int[] {nx, ny}, goal)));
                    cameFrom.put(new int[] {nx, ny}, new int[] {current.x, current.y});
                }
            }
        }

        return null;
    }

    private static List<int[]> reconstructPath(Map<int[], int[]> cameFrom, int[] start, int[] goal) {
        List<int[]> path = new ArrayList<>();
        int[] current = goal;
        while (!Arrays.equals(current, start)) {
            path.add(current);
            current = cameFrom.get(current);
        }
        path.add(start);
        Collections.reverse(path);
        return path;
    }

    private static int heuristic(int[] a, int[] b) {
        return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);
    }

    private static boolean isValid(int[][] grid, int x, int y) {
        return x >= 0 && y >= 0 && x < grid.length && y < grid[0].length && grid[x][y] == 1;
    }

    public static void main(String[] args) {
        int[][] grid = {
            {1, 1, 1, 0},
            {1, 0, 1, 1},
            {1, 1, 1, 0},
            {0, 1, 1, 1}
        };
        int[] start = {0, 0};
        int[] goal = {3, 3};

        List<int[]> path = aStar(grid, start, goal);
        if (path != null) {
            System.out.println("A* Path:");
            for (int[] p : path) System.out.println(Arrays.toString(p));
        } else {
            System.out.println("No path found");
        }
    }
}


// Dijkstra's Algorithm
import java.util.*;

public class DijkstraAlgorithm {
    public static Map<Integer, Integer> dijkstra(Map<Integer, List<int[]>> graph, int start) {
        Map<Integer, Integer> distances = new HashMap<>();
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.add(new int[] {start, 0});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int node = current[0], distance = current[1];

            if (distances.containsKey(node)) continue;
            distances.put(node, distance);

            for (int[] neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (!distances.containsKey(neighbor[0])) {
                    pq.add(new int[] {neighbor[0], distance + neighbor[1]});
                }
            }
        }

        return distances;
    }

    public static void main(String[] args) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(new int[] {1, 4}, new int[] {2, 1}));
        graph.put(1, Arrays.asList(new int[] {2, 2}, new int[] {3, 5}));
        graph.put(2, Arrays.asList(new int[] {3, 8}));
        graph.put(3, Arrays.asList(new int[] {4, 6}));
        graph.put(4, new ArrayList<>());

        Map<Integer, Integer> distances = dijkstra(graph, 0);
        System.out.println("Shortest distances from node 0: " + distances);
    }
}


// Prim's Algorithm
import java.util.*;

public class PrimsAlgorithm {
    public static List<int[]> primsAlgorithm(Map<Integer, List<int[]>> graph, int start) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        Set<Integer> visited = new HashSet<>();
        List<int[]> mst = new ArrayList<>();

        pq.add(new int[] {0, start, -1});

        while (!pq.isEmpty()) {
            int[] edge = pq.poll();
            int weight = edge[0], node = edge[1], parent = edge[2];

            if (visited.contains(node)) continue;
            visited.add(node);

            if (parent != -1) mst.add(new int[] {parent, node, weight});

            for (int[] neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (!visited.contains(neighbor[0])) {
                    pq.add(new int[] {neighbor[1], neighbor[0], node});
                }
            }
        }

        return mst;
    }

    public static void main(String[] args) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(new int[] {1, 2}, new int[] {3, 6}));
        graph.put(1, Arrays.asList(new int[] {0, 2}, new int[] {2, 3}, new int[] {3, 8}));
        graph.put(2, Arrays.asList(new int[] {1, 3}, new int[] {3, 5}));
        graph.put(3, Arrays.asList(new int[] {0, 6}, new int[] {1, 8}, new int[] {2, 5}));

        List<int[]> mst = primsAlgorithm(graph, 0);
        System.out.println("Prim's MST:");
        for (int[] edge : mst) {
            System.out.println(Arrays.toString(edge));
        }
    }
}


// Kruskal's Algorithm
import java.util.*;

public class KruskalsAlgorithm {
    static class Edge implements Comparable<Edge> {
        int src, dest, weight;

        Edge(int src, int dest, int weight) {
            this.src = src;
            this.dest = dest;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge other) {
            return this.weight - other.weight;
        }
    }

    static class UnionFind {
        int[] parent, rank;

        UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            for (int i = 0; i < n; i++) parent[i] = i;
        }

        int find(int x) {
            if (parent[x] != x) parent[x] = find(parent[x]);
            return parent[x];
        }

        void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);

            if (rootX != rootY) {
                if (rank[rootX] > rank[rootY]) {
                    parent[rootY] = rootX;
                } else if (rank[rootX] < rank[rootY]) {
                    parent[rootX] = rootY;
                } else {
                    parent[rootY] = rootX;
                    rank[rootX]++;
                }
            }
        }
    }

    public static List<Edge> kruskalAlgorithm(List<Edge> edges, int n) {
        Collections.sort(edges);
        UnionFind uf = new UnionFind(n);
        List<Edge> mst = new ArrayList<>();

        for (Edge edge : edges) {
            if (uf.find(edge.src) != uf.find(edge.dest)) {
                mst.add(edge);
                uf.union(edge.src, edge.dest);
            }
        }

        return mst;
    }

    public static void main(String[] args) {
        List<Edge> edges = Arrays.asList(
            new Edge(0, 1, 10),
            new Edge(0, 2, 6),
            new Edge(0, 3, 5),
            new Edge(1, 3, 15),
            new Edge(2, 3, 4)
        );
        int n = 4;

        List<Edge> mst = kruskalAlgorithm(edges, n);
        System.out.println("Kruskal's MST:");
        for (Edge edge : mst) {
            System.out.println(edge.src + " - " + edge.dest + " : " + edge.weight);
        }
    }
}


// Eulerian Path/Circuit
import java.util.*;

public class EulerianPath {
    public static List<Integer> findEulerianPath(Map<Integer, List<Integer>> graph, int start) {
        Stack<Integer> stack = new Stack<>();
        List<Integer> path = new ArrayList<>();
        stack.push(start);

        while (!stack.isEmpty()) {
            int u = stack.peek();
            if (graph.get(u) != null && !graph.get(u).isEmpty()) {
                int v = graph.get(u).remove(0);
                stack.push(v);
            } else {
                path.add(stack.pop());
            }
        }

        Collections.reverse(path);
        return path;
    }

    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, new ArrayList<>(Arrays.asList(1, 2)));
        graph.put(1, new ArrayList<>(Arrays.asList(0, 2)));
        graph.put(2, new ArrayList<>(Arrays.asList(0, 1, 3)));
        graph.put(3, new ArrayList<>(Arrays.asList(2)));

        List<Integer> path = findEulerianPath(graph, 0);
        System.out.println("Eulerian Path: " + path);
    }
}



// Hamiltonian Path/Circuit
import java.util.*;

public class HamiltonianPathCircuit {
    public static boolean isSafe(int vertex, boolean[] visited, List<Integer> path, Map<Integer, List<Integer>> graph) {
        return !visited[vertex] && (path.isEmpty() || graph.get(path.get(path.size() - 1)).contains(vertex));
    }

    public static boolean hamiltonianPathUtil(Map<Integer, List<Integer>> graph, boolean[] visited, List<Integer> path, int n) {
        if (path.size() == n) {
            return true;  // All vertices are in the path
        }

        for (int neighbor : graph.getOrDefault(path.get(path.size() - 1), new ArrayList<>())) {
            if (isSafe(neighbor, visited, path, graph)) {
                visited[neighbor] = true;
                path.add(neighbor);

                if (hamiltonianPathUtil(graph, visited, path, n)) return true;

                // Backtrack
                visited[neighbor] = false;
                path.remove(path.size() - 1);
            }
        }

        return false;
    }

    public static List<Integer> findHamiltonianPath(Map<Integer, List<Integer>> graph, int n) {
        for (int start : graph.keySet()) {
            List<Integer> path = new ArrayList<>();
            boolean[] visited = new boolean[n];
            path.add(start);
            visited[start] = true;

            if (hamiltonianPathUtil(graph, visited, path, n)) {
                return path;
            }
        }
        return null;
    }

    public static List<Integer> findHamiltonianCircuit(Map<Integer, List<Integer>> graph, int n) {
        List<Integer> path = findHamiltonianPath(graph, n);
        if (path != null && graph.get(path.get(path.size() - 1)).contains(path.get(0))) {
            path.add(path.get(0));  // Close the circuit
            return path;
        }
        return null;
    }

    public static void main(String[] args) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(1, 2));
        graph.put(1, Arrays.asList(0, 2, 3));
        graph.put(2, Arrays.asList(0, 1, 3));
        graph.put(3, Arrays.asList(1, 2));

        List<Integer> path = findHamiltonianPath(graph, 4);
        if (path != null) {
            System.out.println("Hamiltonian Path: " + path);
        } else {
            System.out.println("No Hamiltonian Path exists");
        }

        List<Integer> circuit = findHamiltonianCircuit(graph, 4);
        if (circuit != null) {
            System.out.println("Hamiltonian Circuit: " + circuit);
        } else {
            System.out.println("No Hamiltonian Circuit exists");
        }
    }
}

File: All DSA C++ Files/python/traversal.py
# In-Order Traversal
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def in_order_traversal(root):
    if root:
        in_order_traversal(root.left)
        print(root.data, end=" ")
        in_order_traversal(root.right)

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    print("In-Order Traversal:", end=" ")
    in_order_traversal(root)
    print()

# Pre-Order Traversal
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def pre_order_traversal(root):
    if root:
        print(root.data, end=" ")
        pre_order_traversal(root.left)
        pre_order_traversal(root.right)

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    print("Pre-Order Traversal:", end=" ")
    pre_order_traversal(root)
    print()


# Post-Order Traversal
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def post_order_traversal(root):
    if root:
        post_order_traversal(root.left)
        post_order_traversal(root.right)
        print(root.data, end=" ")

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    print("Post-Order Traversal:", end=" ")
    post_order_traversal(root)
    print()


# DFS Traversal
def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start, end=" ")
    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs(graph, neighbor, visited)

if __name__ == "__main__":
    graph = {
        0: [1, 2],
        1: [0, 3, 4],
        2: [0, 4],
        3: [1, 5],
        4: [1, 2, 5],
        5: [3, 4]
    }
    print("DFS Traversal:", end=" ")
    dfs(graph, 0)
    print()


# Level-Order Traversal
from collections import deque

class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def level_order_traversal(root):
    if not root:
        return
    queue = deque([root])
    while queue:
        node = queue.popleft()
        print(node.data, end=" ")
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    print("Level-Order Traversal:", end=" ")
    level_order_traversal(root)
    print()


# Backtracking Traversal: N-Queens Problem
def is_safe(board, row, col, n):
    for i in range(col):
        if board[row][i] == 1:
            return False
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
    for i, j in zip(range(row, n), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
    return True

def solve_n_queens(board, col, n):
    if col >= n:
        return True
    for i in range(n):
        if is_safe(board, i, col, n):
            board[i][col] = 1
            if solve_n_queens(board, col + 1, n):
                return True
            board[i][col] = 0
    return False

def print_board(board):
    for row in board:
        print(" ".join("Q" if cell == 1 else "." for cell in row))
    print()

if __name__ == "__main__":
    n = 8
    board = [[0 for _ in range(n)] for _ in range(n)]
    if solve_n_queens(board, 0, n):
        print("N-Queens Solution:")
        print_board(board)
    else:
        print("No solution exists")


# A* Search Algorithm
from heapq import heappop, heappush

def heuristic(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])  # Manhattan distance

def a_star_search(grid, start, goal):
    rows, cols = len(grid), len(grid[0])
    open_set = []
    heappush(open_set, (0 + heuristic(start, goal), 0, start))
    came_from = {}
    g_score = {start: 0}

    while open_set:
        _, current_cost, current = heappop(open_set)

        if current == goal:
            path = []
            while current in came_from:
                path.append(current)
                current = came_from[current]
            path.append(start)
            return path[::-1]

        for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            neighbor = (current[0] + dx, current[1] + dy)
            if 0 <= neighbor[0] < rows and 0 <= neighbor[1] < cols and grid[neighbor[0]][neighbor[1]] == 1:
                tentative_g_score = current_cost + 1
                if tentative_g_score < g_score.get(neighbor, float('inf')):
                    g_score[neighbor] = tentative_g_score
                    heappush(open_set, (tentative_g_score + heuristic(neighbor, goal), tentative_g_score, neighbor))
                    came_from[neighbor] = current
    return None

if __name__ == "__main__":
    grid = [
        [1, 1, 1, 0],
        [1, 0, 1, 1],
        [1, 1, 1, 0],
        [0, 1, 1, 1]
    ]
    start = (0, 0)
    goal = (3, 3)

    path = a_star_search(grid, start, goal)
    if path:
        print("A* Path:", path)
    else:
        print("No path found")


# Dijkstra's Algorithm
from heapq import heappop, heappush

def dijkstra(graph, start):
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    priority_queue = [(0, start)]
    visited = set()

    while priority_queue:
        current_distance, current_node = heappop(priority_queue)

        if current_node in visited:
            continue
        visited.add(current_node)

        for neighbor, weight in graph[current_node]:
            distance = current_distance + weight
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heappush(priority_queue, (distance, neighbor))

    return distances

if __name__ == "__main__":
    graph = {
        0: [(1, 4), (2, 1)],
        1: [(2, 2), (3, 5)],
        2: [(3, 8)],
        3: [(4, 6)],
        4: []
    }

    distances = dijkstra(graph, 0)
    print("Shortest distances from node 0:", distances)


# Prim's Algorithm
from heapq import heappop, heappush

def prims_algorithm(graph):
    mst = []
    visited = set()
    edges = [(0, 0, 0)]  # (weight, start, end)

    while edges:
        weight, start, end = heappop(edges)
        if end not in visited:
            visited.add(end)
            mst.append((start, end, weight))

            for neighbor, cost in graph[end]:
                if neighbor not in visited:
                    heappush(edges, (cost, end, neighbor))

    return mst

if __name__ == "__main__":
    graph = {
        0: [(1, 2), (3, 6)],
        1: [(0, 2), (2, 3), (3, 8)],
        2: [(1, 3), (3, 5)],
        3: [(0, 6), (1, 8), (2, 5)]
    }

    mst = prims_algorithm(graph)
    print("Prim's MST:", mst)


# Kruskal's Algorithm
def find(parent, node):
    if parent[node] != node:
        parent[node] = find(parent, parent[node])
    return parent[node]

def union(parent, rank, u, v):
    root_u = find(parent, u)
    root_v = find(parent, v)
    if root_u != root_v:
        if rank[root_u] > rank[root_v]:
            parent[root_v] = root_u
        elif rank[root_u] < rank[root_v]:
            parent[root_u] = root_v
        else:
            parent[root_v] = root_u
            rank[root_u] += 1

def kruskals_algorithm(edges, n):
    edges.sort(key=lambda x: x[2])
    parent = list(range(n))
    rank = [0] * n
    mst = []

    for u, v, weight in edges:
        if find(parent, u) != find(parent, v):
            mst.append((u, v, weight))
            union(parent, rank, u, v)

    return mst

if __name__ == "__main__":
    edges = [
        (0, 1, 10), (0, 2, 6), (0, 3, 5),
        (1, 3, 15), (2, 3, 4)
    ]
    n = 4
    mst = kruskals_algorithm(edges, n)
    print("Kruskal's MST:", mst)


# Eulerian Path/Circuit
from collections import defaultdict

def find_eulerian_path(graph, start):
    path = []
    stack = [start]

    while stack:
        u = stack[-1]
        if graph[u]:
            v = graph[u].pop()
            stack.append(v)
        else:
            path.append(stack.pop())

    return path[::-1]

if __name__ == "__main__":
    graph = defaultdict(list, {
        0: [1, 2],
        1: [0, 2],
        2: [0, 1, 3],
        3: [2]
    })
    path = find_eulerian_path(graph, 0)
    print("Eulerian Path:", path)


# Hamiltonian Path
def is_hamiltonian_path(graph, path):
    n = len(graph)
    if len(path) != n:
        return False
    visited = set()
    for i in range(n):
        if path[i] in visited or (i > 0 and path[i] not in graph[path[i - 1]]):
            return False
        visited.add(path[i])
    return True

def hamiltonian_path_util(graph, path, visited):
    if len(path) == len(graph):  # Base case: All vertices are in the path
        return True

    for neighbor in graph[path[-1]]:
        if neighbor not in visited:
            visited.add(neighbor)
            path.append(neighbor)

            if hamiltonian_path_util(graph, path, visited):
                return True

            # Backtrack
            path.pop()
            visited.remove(neighbor)

    return False

def find_hamiltonian_path(graph):
    for start in graph:
        path = [start]
        visited = {start}

        if hamiltonian_path_util(graph, path, visited):
            return path

    return None

if __name__ == "__main__":
    graph = {
        0: [1, 2],
        1: [0, 2, 3],
        2: [0, 1, 3],
        3: [1, 2]
    }

    path = find_hamiltonian_path(graph)
    if path:
        print("Hamiltonian Path:", path)
    else:
        print("No Hamiltonian Path exists")


# Hamiltonian Circuit
def is_hamiltonian_circuit(graph, path):
    n = len(graph)
    if len(path) != n or path[0] not in graph[path[-1]]:
        return False
    visited = set()
    for i in range(n):
        if path[i] in visited or (i > 0 and path[i] not in graph[path[i - 1]]):
            return False
        visited.add(path[i])
    return True

def hamiltonian_circuit_util(graph, path, visited):
    if len(path) == len(graph):
        return path[0] in graph[path[-1]]  # Check if circuit exists

    for neighbor in graph[path[-1]]:
        if neighbor not in visited:
            visited.add(neighbor)
            path.append(neighbor)

            if hamiltonian_circuit_util(graph, path, visited):
                return True

            # Backtrack
            path.pop()
            visited.remove(neighbor)

    return False

def find_hamiltonian_circuit(graph):
    for start in graph:
        path = [start]
        visited = {start}

        if hamiltonian_circuit_util(graph, path, visited):
            path.append(path[0])  # Close the circuit
            return path

    return None

if __name__ == "__main__":
    graph = {
        0: [1, 2],
        1: [0, 2, 3],
        2: [0, 1, 3],
        3: [1, 2]
    }

    circuit = find_hamiltonian_circuit(graph)
    if circuit:
        print("Hamiltonian Circuit:", circuit)
    else:
        print("No Hamiltonian Circuit exists")
